File: backend/src/main/java/com/aura_card/backend/controller/UserController.java
package com.aura_card.backend.controller;

import com.aura_card.backend.dto.request.CreateUserRequest;
import com.aura_card.backend.dto.request.UpdateUserRequest;
import com.aura_card.backend.dto.response.AccountResponse;
import com.aura_card.backend.dto.response.ApiResponse;
import com.aura_card.backend.dto.response.UserResponse;
import com.aura_card.backend.service.UserService;
import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/api/users")
public class UserController {

    private final UserService userService;

    public UserController(UserService userService) {
        this.userService = userService;
    }

    // Lấy danh sách users
    @GetMapping
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ApiResponse<List<UserResponse>>> getUsers(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "12") int limit,
            @RequestParam(required = false) String q,
            @RequestParam(required = false) String role,
            @RequestParam(required = false) String status) {

        Page<UserResponse> data = userService.getUsers(page, limit, q, role, status);

        return ResponseEntity.ok(ApiResponse.<List<UserResponse>>builder()
                .message("Lấy danh sách người dùng thành công")
                .data(data.getContent())
                .total(data.getTotalElements())
                .totalPages(data.getTotalPages())
                .build());
    }

    // Lấy user theo id
    @GetMapping("/{userId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ApiResponse<UserResponse>> getUserById(@PathVariable String userId) {
        return ResponseEntity.ok(ApiResponse.<UserResponse>builder()
                .message("Lấy người dùng thành công")
                .data(userService.getUserById(userId))
                .build());
    }

    // Lấy tài khoản đang đăng nhập
    @GetMapping("/me")
    @PreAuthorize("hasAnyRole('ADMIN', 'CUSTOMER')")
    public ResponseEntity<ApiResponse<AccountResponse>> getAccount(
            @AuthenticationPrincipal String userId) {
        return ResponseEntity.ok(ApiResponse.<AccountResponse>builder()
                .message("Lấy tài khoản thành công")
                .data(userService.getAccount(userId))
                .build());
    }

    // Tạo người dùng
    @PostMapping
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ApiResponse<UserResponse>> createUser(@RequestBody @Valid CreateUserRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.<UserResponse>builder()
                        .message("Tạo người dùng thành công")
                        .data(userService.createUser(request))
                        .build());
    }

    // Cập nhật người dùng
    @PutMapping("/{userId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ApiResponse<UserResponse>> updateUser(
            @PathVariable String userId,
            @RequestBody @Valid UpdateUserRequest request) {
        return ResponseEntity.ok(ApiResponse.<UserResponse>builder()
                .message("Cập nhật người dùng thành công")
                .data(userService.updateUser(userId, request))
                .build());
    }

    // Cập nhật status ACTIVE <-> LOCKED
    @PatchMapping("/status/{userId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ApiResponse<UserResponse>> UpdateUserStatus(@PathVariable String userId) {
        return ResponseEntity.ok(ApiResponse.<UserResponse>builder()
                .message("Cập nhật tình trạng thành công")
                .data(userService.updateUserStatus(userId))
                .build());
    }

    // Xóa người dùng
    @DeleteMapping("/{userId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ApiResponse<Void>> deleteUser(@PathVariable String userId) {
        userService.deleteUser(userId);
        return ResponseEntity.ok(ApiResponse.<Void>builder()
                .message("Xóa người dùng thành công")
                .build());
    }
}

File: backend-nodejs/controllers/user.controller.js
const userService = require("../services/user.service");
const response = require("../utils/response.util");

// Lấy danh sách users
const getUsers = async (req, res, next) => {
  try {
    const { page, limit, q, role, status } = req.query;
    const result = await userService.getUsers({ page, limit, q, role, status });
    response.success(res, {
      message: "Lấy danh sách người dùng thành công",
      data: result.data,
      total: result.total,
      totalPages: result.totalPages,
    });
  } catch (error) {
    next(error);
  }
};

// Lấy user theo id
const getUserById = async (req, res, next) => {
  try {
    const data = await userService.getUserById(req.params.userId);
    response.success(res, { message: "Lấy người dùng thành công", data });
  } catch (error) {
    next(error);
  }
};

// Lấy tài khoản đang đăng nhập
const getAccount = async (req, res, next) => {
  try {
    const data = await userService.getAccount(req.user.id);
    response.success(res, { message: "Lấy tài khoản thành công", data });
  } catch (error) {
    next(error);
  }
};

// Tạo người dùng
const createUser = async (req, res, next) => {
  try {
    const data = await userService.createUser(req.body);
    response.success(res, {
      message: "Tạo người dùng thành công",
      data,
      status: 201,
    });
  } catch (error) {
    next(error);
  }
};

// Cập nhật người dùng
const updateUser = async (req, res, next) => {
  try {
    const data = await userService.updateUser(req.params.userId, req.body);
    response.success(res, { message: "Cập nhật người dùng thành công", data });
  } catch (error) {
    next(error);
  }
};

// Cập nhật status
const updateUserStatus = async (req, res, next) => {
  try {
    const data = await userService.updateUserStatus(req.params.userId);
    response.success(res, { message: "Cập nhật tình trạng thành công", data });
  } catch (error) {
    next(error);
  }
};

// Xóa người dùng
const deleteUser = async (req, res, next) => {
  try {
    await userService.deleteUser(req.params.userId);
    response.success(res, { message: "Xóa người dùng thành công" });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  getUsers,
  getUserById,
  getAccount,
  createUser,
  updateUser,
  updateUserStatus,
  deleteUser,
};
